Fixes #37900 - Allow syncing templates through HTTP proxy

adamlazik1 opened the pull request.

adamlazik1 commented:
TODO:
- tests
- only display HTTP proxy selection box if the right proxy policy is selected

lhellebr commented:
I've just tested the sweet paths and can confirm API part works.

lhellebr commented:
OK, even WebUI part seems to work. Is my testing wrong or is that expected? My last info was that this should only work in API.

adamlazik1 commented:
@lhellebr What I currently don't like about the web ui is that the the HTTP proxy selection box appears even when you have global proxy or no proxy selected in the proxy policy box. This could potentially confuse the user. I am currently trying to fix that. The behavior where the selection box appears only when the proxy policy is "selected"  would be also consistent with for example the one from katello - on the **Content -> Products -> Repositories -> New repository** page, where they handle HTTP proxies the same way. Unfortunately the front end there uses different technology so I cannot copy the code. I am very close to the solution, though.

adamlazik1 commented:
Alright, now the proxy id field is displayed only when it's supposed to but I've broken the ui procedure as proxy params are currently not sent to the controller. I also need to figure out how to make translations work on the new fields. On the plus side, username and password for a proxy is now used correctly. I still need to setup config for SSL cert. I'll have to continue on this next week.

adamlazik1 commented:
Right now everything works if there is at least one proxy to choose but things break in the UI if there is no proxy available. I will have to fix that next week.

adamlazik1 commented on webpack/components/NewTemplateSync/components/NewTemplateSyncForm/NewTemplateSyncFormHelpers.js:
```diff
@@ -1,5 +1,8 @@
 import * as Yup from 'yup';
 
+import React from 'react';
```
Without it I am getting the below error so I suspect maybe in the `tooltipContent` function.
```
Uncaught ReferenceError: React is not defined
    tooltipContent NewTemplateSyncFormHelpers.js:46
    SyncSettingField SyncSettingField.js:36
    React
    unstable_runWithPriority scheduler.development.js:653
    React 4
    Redux 4
    routerMiddleware middleware.js:22
    Redux 4
    _callee$ NewTemplateSyncActions.js:22
    tryCatch NewTemplateSyncActions.js:2
    makeInvokeMethod NewTemplateSyncActions.js:2
    defineIteratorMethods NewTemplateSyncActions.js:2
    asyncGeneratorStep NewTemplateSyncActions.js:2
    _next NewTemplateSyncActions.js:2
9 [NewTemplateSyncFormHelpers.js:46:37](webpack://foreman_templates/components/NewTemplateSync/components/NewTemplateSyncForm/NewTemplateSyncFormHelpers.js)
    tooltipContent NewTemplateSyncFormHelpers.js:46
    SyncSettingField SyncSettingField.js:36
    React 11
    unstable_runWithPriority scheduler.development.js:653
    React 4
    Redux 4
    routerMiddleware middleware.js:22
    Redux 4
    _callee$ NewTemplateSyncActions.js:22
    tryCatch NewTemplateSyncActions.js:2
    makeInvokeMethod NewTemplateSyncActions.js:2
    defineIteratorMethods NewTemplateSyncActions.js:2
    asyncGeneratorStep NewTemplateSyncActions.js:2
    _next NewTemplateSyncActions.js:2
```

adamlazik1 commented on app/services/foreman_templates/action.rb:
```diff
@@ -53,9 +55,43 @@ def verify_path!(path)
+        http_proxy = HttpProxy.authorized(:view_http_proxies).with_taxonomy_scope.find(@http_proxy_id)
+        http_proxy_url = http_proxy.full_url
+
+        if http_proxy_url.include?('https') && http_proxy.cacert.present?
```
Applied.

adamlazik1 commented on app/services/foreman_templates/action.rb:
```diff
@@ -53,9 +55,43 @@ def verify_path!(path)
+        http_proxy_url = http_proxy.full_url
+
+        if http_proxy_url.include?('https') && http_proxy.cacert.present?
+          @proxy_cert = "/tmp/foreman_templates_proxy_cert_#{SecureRandom.hex(8)}.crt"
```
Not with the new proxy cert location, fixed.

adamlazik1 commented on app/services/foreman_templates/action.rb:
```diff
@@ -53,9 +55,43 @@ def verify_path!(path)
+        http_proxy_url = http_proxy.full_url
+
+        if http_proxy_url.include?('https') && http_proxy.cacert.present?
+          @proxy_cert = "/tmp/foreman_templates_proxy_cert_#{SecureRandom.hex(8)}.crt"
```
Applied.

asteflova commented on app/controllers/api/v2/template_controller.rb:
```diff
@@ -12,7 +12,9 @@ class TemplateController < ::Api::V2::BaseController
         param :negate, :bool, :required => false, :desc => N_("Negate the prefix (for purging).")
-        param :dirname, String, :required => false, :desc => N_("The directory within Git repo containing the templates")
+        param :dirname, String, :required => false, :desc => N_("Directory within Git repo containing the templates.")
+        param :http_proxy_policy, ForemanTemplates.http_proxy_policy_types.keys, :required => false, :desc => N_("HTTP proxy policy for template sync.")
```
```suggestion
        param :http_proxy_policy, ForemanTemplates.http_proxy_policy_types.keys, :required => false, :desc => N_("HTTP proxy policy for Template Sync.")
```

asteflova commented on app/controllers/api/v2/template_controller.rb:
```diff
@@ -12,7 +12,9 @@ class TemplateController < ::Api::V2::BaseController
-        param :dirname, String, :required => false, :desc => N_("The directory within Git repo containing the templates")
+        param :dirname, String, :required => false, :desc => N_("Directory within Git repo containing the templates.")
+        param :http_proxy_policy, ForemanTemplates.http_proxy_policy_types.keys, :required => false, :desc => N_("HTTP proxy policy for template sync.")
+        param :http_proxy_id, :number, :required => false, :desc => N_("ID of an HTTP proxy to use for template sync.")
```
```suggestion
        param :http_proxy_id, :number, :required => false, :desc => N_("ID of an HTTP proxy to use for Template Sync.")
```

asteflova reviewed (commented):
I took a peek at the new settings in the web UI and left a few comments about the things that I expect might be unclear, keeping in mind that whatever is unclear on the UI side would need to be explained in documentation :)

adamlazik1 commented:
Fixed the last issue that I found and simplified the code a bit. From the functionality side this should be fully functional. I still need to add tests.

asteflova commented on lib/foreman_templates/engine.rb:
```diff
@@ -94,6 +94,12 @@ class Engine < ::Rails::Engine
               full_name: N_('Commit message'))
+            setting('template_sync_http_proxy_policy',
+              type: :string,
+              description: N_('HTTP proxy policy for template sync'),
```
> About describing the options: Ideally I want to reach state where options themselves would be descriptive enough.

:+1: :100: 

> Probably the most unclear is "Custom HTTP proxy". Do you think it would be okay to is provide more detail for that one and omitting the other two in the description?

Yes! I like this idea. How about just adding a few words to explain that if users select "Custom HTTP proxy", they will be prompted for the proxy (because after you select `Custom`, a new field appears). Together with the new field now including the hint about "Infrastructure > HTTP proxies", users shouldn't get lost -- and you avoid overdescribing the actual options. WDYT?

> I wonder if capitalizing template sync is a good idea. Is this to keep consistency with a different part of the UI?

Yes, that was the idea. But now that I'm looking again, is Template Sync capitalized only when it's used as a heading..? Okay, please disregard these comments :) We don't list is as a term to be capitalized in the docs [CONTRIBUTING.md](https://github.com/theforeman/foreman-documentation/blob/master/CONTRIBUTING.md#capitalization) so let's not do it here either.

asteflova replied:
Work for me! Now we won't have to explain all of this separately on the docs side. Thanks!

lhellebr commented:
@adamlazik1 
```
# hammer -v import-templates  --repo="https://github.com/theforeman/community-templates.git" --organization-ids="36" --branch="develop" --dirname="/partition_tables_templates/" --filter="FreeBSD default fake" --http-proxy-policy="none"
Could not import:
  Error: undefined method `empty?' for nil:NilClass
# hammer  import-templates  --repo="https://github.com/theforeman/community-templates.git" --organization-ids="36" --branch="develop" --dirname="/partition_tables_templates/" --filter="FreeBSD default fake" --http-proxy-policy="none"
Import finished.
```

lhellebr commented:
In WebUI, when syncing with default form values, just specifying branch to `develop` (and any setting of proxy), on stream, everything passes. With this PR applied, however, I get an empty page (except menu) and an error notification:
```
Error
Oh no! Something went wrong while submitting the form, the server returned the following error: 
```
The templates still seem to be applied. `foreman-tail` doesn't seem to be very useful:
```
2024-11-05T08:55:13 [I|app|645ea88a] Started POST "/ui_template_syncs/import" for <IP> at 2024-11-05 08:55:13 -0500
2024-11-05T08:55:13 [I|app|645ea88a] Processing by UITemplateSyncsController#import as JSON
2024-11-05T08:55:13 [I|app|645ea88a]   Parameters: {"repo"=>"https://github.com/theforeman/community-templates.git", "branch"=>"develop", "dirname"=>"/", "filter"=>"", "negate"=>false, "prefix"=>"", "associate"=>"new", "force"=>false, "lock"=>"keep", "http_proxy_policy"=>"global", "http_proxy_id"=>1, "ui_template_sync"=>{"repo"=>"https://github.com/theforeman/community-templates.git", "branch"=>"develop", "dirname"=>"/", "filter"=>"", "negate"=>false, "prefix"=>"", "associate"=>"new", "force"=>false, "lock"=>"keep", "http_proxy_policy"=>"global", "http_proxy_id"=>1}}
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.

==> /var/log/httpd/foreman-ssl_access_ssl.log <==
10.39.195.74 - - [05/Nov/2024:08:55:13 -0500] "POST /ui_template_syncs/import HTTP/2.0" 200 57999 "https://<FQDN>/template_syncs" "Mozilla/5.0 (X11; Linux x86_64; rv:132.0) Gecko/20100101 Firefox/132.0"

==> /var/log/foreman/production.log <==
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.

[...]

2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
2024-11-05T08:55:14 [I|app|645ea88a]   Rendered /usr/share/gems/gems/foreman_templates-10.0.0/app/views/ui_template_syncs/import.json.rabl (Duration: 104.3ms | Allocations: 111180)
2024-11-05T08:55:14 [I|app|645ea88a] Completed 200 OK in 1509ms (Views: 105.0ms | ActiveRecord: 169.0ms | Allocations: 633230)

```

adamlazik1 commented:
> In WebUI, when syncing with default form values, just specifying branch to `develop` (and any setting of proxy), on stream, everything passes. With this PR applied, however, I get an empty page (except menu) and an error notification:
> 
> ```
> Error
> Oh no! Something went wrong while submitting the form, the server returned the following error: 
> ```
> 
> The templates still seem to be applied. `foreman-tail` doesn't seem to be very useful:
> 
> ```
> 2024-11-05T08:55:13 [I|app|645ea88a] Started POST "/ui_template_syncs/import" for <IP> at 2024-11-05 08:55:13 -0500
> 2024-11-05T08:55:13 [I|app|645ea88a] Processing by UITemplateSyncsController#import as JSON
> 2024-11-05T08:55:13 [I|app|645ea88a]   Parameters: {"repo"=>"https://github.com/theforeman/community-templates.git", "branch"=>"develop", "dirname"=>"/", "filter"=>"", "negate"=>false, "prefix"=>"", "associate"=>"new", "force"=>false, "lock"=>"keep", "http_proxy_policy"=>"global", "http_proxy_id"=>1, "ui_template_sync"=>{"repo"=>"https://github.com/theforeman/community-templates.git", "branch"=>"develop", "dirname"=>"/", "filter"=>"", "negate"=>false, "prefix"=>"", "associate"=>"new", "force"=>false, "lock"=>"keep", "http_proxy_policy"=>"global", "http_proxy_id"=>1}}
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 
> ==> /var/log/httpd/foreman-ssl_access_ssl.log <==
> 10.39.195.74 - - [05/Nov/2024:08:55:13 -0500] "POST /ui_template_syncs/import HTTP/2.0" 200 57999 "https://<FQDN>/template_syncs" "Mozilla/5.0 (X11; Linux x86_64; rv:132.0) Gecko/20100101 Firefox/132.0"
> 
> ==> /var/log/foreman/production.log <==
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 
> [...]
> 
> 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
> 2024-11-05T08:55:14 [I|app|645ea88a]   Rendered /usr/share/gems/gems/foreman_templates-10.0.0/app/views/ui_template_syncs/import.json.rabl (Duration: 104.3ms | Allocations: 111180)
> 2024-11-05T08:55:14 [I|app|645ea88a] Completed 200 OK in 1509ms (Views: 105.0ms | ActiveRecord: 169.0ms | Allocations: 633230)
> ```
> 
> The messages about template being locked are shown in stream without this PR as well.

I tried this on the master branch and got the same result, it seems that it is unrelated to this PR. I will try to find the cause.

adamlazik1 commented:
The unit tests here should hopefully be enough. I don't know if I should create any API controller tests because I can't simulate an HTTP proxy and can't think about testing anything else that would make sense. If someone has any ideas, I can add more tests, but until then I consider the added tests sufficient.

adamlazik1 commented:
> > In WebUI, when syncing with default form values, just specifying branch to `develop` (and any setting of proxy), on stream, everything passes. With this PR applied, however, I get an empty page (except menu) and an error notification:
> > ```
> > Error
> > Oh no! Something went wrong while submitting the form, the server returned the following error: 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The templates still seem to be applied. `foreman-tail` doesn't seem to be very useful:
> > ```
> > 2024-11-05T08:55:13 [I|app|645ea88a] Started POST "/ui_template_syncs/import" for <IP> at 2024-11-05 08:55:13 -0500
> > 2024-11-05T08:55:13 [I|app|645ea88a] Processing by UITemplateSyncsController#import as JSON
> > 2024-11-05T08:55:13 [I|app|645ea88a]   Parameters: {"repo"=>"https://github.com/theforeman/community-templates.git", "branch"=>"develop", "dirname"=>"/", "filter"=>"", "negate"=>false, "prefix"=>"", "associate"=>"new", "force"=>false, "lock"=>"keep", "http_proxy_policy"=>"global", "http_proxy_id"=>1, "ui_template_sync"=>{"repo"=>"https://github.com/theforeman/community-templates.git", "branch"=>"develop", "dirname"=>"/", "filter"=>"", "negate"=>false, "prefix"=>"", "associate"=>"new", "force"=>false, "lock"=>"keep", "http_proxy_policy"=>"global", "http_proxy_id"=>1}}
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 
> > ==> /var/log/httpd/foreman-ssl_access_ssl.log <==
> > 10.39.195.74 - - [05/Nov/2024:08:55:13 -0500] "POST /ui_template_syncs/import HTTP/2.0" 200 57999 "https://<FQDN>/template_syncs" "Mozilla/5.0 (X11; Linux x86_64; rv:132.0) Gecko/20100101 Firefox/132.0"
> > 
> > ==> /var/log/foreman/production.log <==
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:13 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 
> > [...]
> > 
> > 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:14 [E|app|645ea88a] couldn't save the template because of: Base This template is locked. Please clone it to a new template to customize., This template is locked. Please clone it to a new template to customize.
> > 2024-11-05T08:55:14 [I|app|645ea88a]   Rendered /usr/share/gems/gems/foreman_templates-10.0.0/app/views/ui_template_syncs/import.json.rabl (Duration: 104.3ms | Allocations: 111180)
> > 2024-11-05T08:55:14 [I|app|645ea88a] Completed 200 OK in 1509ms (Views: 105.0ms | ActiveRecord: 169.0ms | Allocations: 633230)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The messages about template being locked are shown in stream without this PR as well.
> 
> I tried this on the master branch and got the same result, it seems that it is unrelated to this PR. I will try to find the cause.

https://github.com/theforeman/foreman_templates/pull/192

lhellebr commented:
2 More issues (with a few days old version, maybe you fixed it already, let me know)

1)
When I select `Use selected HTTP proxy` policy and there is no HTTP proxy created, the field to select the proxy doesn't show, instead it says there are no HTTP proxies. Yet, I can still click submit and I get this not really helpful message that wouldn't immediately point me to proxy if I wasn't testing proxy:
```
Warning! Error submitting data: 404 
```

2)
In Settings -> Template sync, I can set policy but there is no field to set the chosen proxy if I choose `Use selected HTTP proxy`

adamlazik1 commented:
@lhellebr 

> When I select `Use selected HTTP proxy` policy and there is no HTTP proxy created, the field to select the proxy doesn't show, instead it says there are no HTTP proxies. Yet, I can still click submit and I get this not really helpful message that wouldn't immediately point me to proxy if I wasn't testing proxy:

This should be fixed in the current version, the submit button is not clickable if there is no proxy to select. 

> In Settings -> Template sync, I can set policy but there is no field to set the chosen proxy if I choose `Use selected HTTP proxy`

This is intentional, since HTTP proxies are scoped by taxonomies, whereas settings are global. Adding HTTP proxy to settings would mean it would be visible outside of its assigned organizations and locations, which I don't think is correct.

lhellebr commented:
Understood!

adamruzicka reviewed (commented):
The ruby part looks sensible, @MariaAga could you please take a look at the frontend part?

MariaAga commented on webpack/components/NewTemplateSync/components/ProxySettingFields.js:
```diff
@@ -0,0 +1,116 @@
+  const proxyPolicyFieldName = `${syncType}.http_proxy_policy`;
+  const proxyIdFieldName = `${syncType}.http_proxy_id`;
+
+  const descriptionEdits = {
+    '? ': '?<br>',
+    'Custom HTTP proxy': '<b>Custom HTTP proxy</b>',
+    ' when syncing templates': '',
+  };
+  proxyPolicySetting = proxyPolicySetting.set(
+    'description',
+    `${proxyPolicySetting.description.replace(
+      /\b(\? |Custom HTTP proxy| when syncing templates)\b/g,
+      match => descriptionEdits[match]
+    )}`
+  );
```
This causes the description not to be translated.
`N_('Should an HTTP proxy be used for template sync? If you select Custom HTTP proxy, you will be prompted to select one when syncing templates.')` - this marks it to be translated in the future, for example in the js code with `__(proxyPolicySetting.description)`.
Maybe it will be better to use the <br> in lib/foreman_templates/engine.rb but I'm not sure if the description is used anywhere else

MariaAga replied:
1. yes, I didnt check the code before, but it looks like modifyDescription will not be translated.
2. We can add new strings in the UI, that mark for translated is used so we can send variables to be translated:
`translate_later = N('string')` and then `__(translate_later)`

adamlazik1 replied:
I see, thank you for the explanation. In that case, I will not modify the proxy settings descriptions in the UI as I feel like it would unnecessarily complicate the code.

MariaAga commented on webpack/components/NewTemplateSync/components/NewTemplateSyncForm/NewTemplateSyncFormHelpers.js:
```diff
@@ -24,14 +29,30 @@ export const syncFormSchema = (syncType, settingsObj, validationData) => {
+      if (setting.name === 'http_proxy_policy') {
+        return {
+          ...memo,
+          http_proxy_policy: Yup.mixed().test(
+            'http-proxy-available',
+            __(
+              'No HTTP proxies available. Please select a different policy or verify that you have selected the correct organization and location.'
+            ),
```
This is only visible after clicking "No HTTP proxies available" and also blocks the user form changing from import to export. I think it might be better to not show the option for a custom http proxy is the user doesnt have any (or, more complicated I think, to show the warning and only block submit)

adamruzicka reviewed (approved):
Looks good, let's get this in

adamruzicka commented:
Thank you everyone!

lhellebr commented:
Can I expect this in stream snap 81?
